org.adempiere.ui.zk/WEB-INF/src/web/theme/default/css/fragment/gadget.css.dsp

```diff
@@ -267,3 +267,113 @@
   vertical-align: middle;
 }
 <%-- User Favorite Dashboard Panel End --%>
+
+.z-anchorchildren .z-hlayout {
+	white-space: normal;
+}
+
+.dashboard-row-10 .z-hlayout-inner {
+	width: 10%;
+}
+.dashboard-row-9 .z-hlayout-inner {
+	width: 11%;
+}
+.dashboard-row-8 .z-hlayout-inner {
+	width: 12%;
+}
+.dashboard-row-7 .z-hlayout-inner {
+	width: 14%;
+}
+.dashboard-row-6 .z-hlayout-inner {
+	width: 16%;
+}
+.dashboard-row-5 .z-hlayout-inner {
+	width: 20%;
+}
+.dashboard-row-4 .z-hlayout-inner {
+	width: 25%;
+}
+.dashboard-row-3 .z-hlayout-inner {
+	width: 33%;
+}
+.dashboard-row-2 .z-hlayout-inner {
+	width: 50%;
+}
+.dashboard-row-1 .z-hlayout-inner {
+	width: 100%;
+}
+@media screen and (max-width: 1024px) {
+	.dashboard-row-10 .z-hlayout-inner {
+		width: 20%;
+	}
+	.dashboard-row-9 .z-hlayout-inner {
+		width: 25%;
+	}
+	.dashboard-row-9 .z-hlayout-inner:last-child {
+		width: 100%;
+	}
+	.dashboard-row-8 .z-hlayout-inner {
+		width: 25%;
+	}
+	.dashboard-row-7 .z-hlayout-inner {
+		width: 33%;
+	}
+	.dashboard-row-7 .z-hlayout-inner:last-child {
+		width: 100%;
+	}
+	.dashboard-row-6 .z-hlayout-inner {
+		width: 33%;
+	}
+	.dashboard-row-5 .z-hlayout-inner {
+		width: 50%;
+	}
+	.dashboard-row-5 .z-hlayout-inner:last-child {
+		width: 100%;
+	}
+	.dashboard-row-4 .z-hlayout-inner {
+		width: 50%;
+	}
+	.dashboard-row-3 .z-hlayout-inner {
+		width: 50%;
+	}
+	.dashboard-row-3 .z-hlayout-inner:last-child {
+		width: 100%;
+	}
+	.dashboard-row-2 .z-hlayout-inner {
+		width: 100%;
+	}
+}
+@media screen and (max-width: 750px) {
+	.dashboard-row-10 .z-hlayout-inner {
+		width: 100%;
+		display: flex;
+	}
+	.dashboard-row-9 .z-hlayout-inner {
+		width: 100%;
+		display: flex;
+	}
+	.dashboard-row-8 .z-hlayout-inner {
+		width: 100%;
+		display: flex;
+	}
+	.dashboard-row-7 .z-hlayout-inner {
+		width: 100%;
+		display: flex;
+	}
+	.dashboard-row-6 .z-hlayout-inner {
+		width: 100%;
+		display: flex;
+	}
+	.dashboard-row-5 .z-hlayout-inner {
+		width: 100%;
+		display: flex;
+	}
+	.dashboard-row-4 .z-hlayout-inner {
+		width: 100%;
+		display: flex;
+	}
+	.dashboard-row-3 .z-hlayout-inner {
+		width: 100%;
+		display: flex;
+	}
+}
```

Comment:
This basically a hard coded layout. Wouldn't works well for the generic need of the project and it is not going to works fine if the dashboard still support dynamic reordering of dashboard widget by user.

Reply:
Do you mean by hardcoded that we've defined 10 classes (support 10 dashboards in 1 row)? If we would change it to 12, it would be the exact limitation Bootstrap has. We assumed the admin never defines more than 10 gadgets in 1 row.

According to my tests the dynamic reordering of dashboard widget by user worked for me. Probably we need to specify the use case, when does it not work fine?

We focus to keep same principles on server side, and on client side we achieve the better responsiveness by client-side css.

At the same time we think, that the css was not written the best way. If we change the exact number of defined classes to dynamic approach, would this solve the problem? 

For clarifying - does the java improvement seems to be good? Or you haven't checked it yet? 

Could you explain more for resolving the issues, please?

Reply:
Hi @hengsin , we are consulted here the css question with our web designer. We modify the design which also covering span.

Firstly we define a new sysconfig - define "dashboard (row/col) max gadgets"  . let say we set typical boostrap 12 figure. Dashboard Preference will have "GadgetSpan" field, where the admin enter the number which specify how many "positions" cols this gadget taken.  let say set 6 then that means in case 12 max gadgets 50% and additional 2 will have 3+3  (I assume this  can be implemented for column based layout too probably also with support without specify span field)

We will rename css classes, because we had wrong approach - however 12 classes for 12 cols must be defined. + 12 for media small device. (this because we are able client side redraw the layout). we are can't find better solution. 

Another options would be flexbox https://css-tricks.com/snippets/css/a-guide-to-flexbox/  or replace fully grid by javascript lib which is for flexible layout generation. (the second is out of our knowledge for now.)

if you have another idea let us know.